codeU2.py

```diff
@@ -0,0 +1,210 @@
+# -*- coding: utf-8 -*-
+"""
+Created on Fri Jan 13 10:22:28 2017
+Exercise set 2, Code U
+@author: Luiza
+"""
+import logging
+import unittest
+
+def dfs(tree, start):
+
+    """Depth-first search
+    """
+
+    if tree is None:
+        return
+    visited = set()
+    stack = [start]
+    while stack:
+        vertex = stack.pop()
+        if not vertex:
+            continue
+        visited.add(vertex)
+        stack.extend(set(tree[vertex]) - visited)
+    return visited
+
+def dfs_path(tree, start, end, path=None):
+
+    """Returns all full paths from start node
+    to end node, including the end node
+    """
+
+    stack = [(start, [start])]
+    while stack:
+        (vertex, path) = stack.pop()
+        if vertex is None:
+            continue
+        for next in set(tree[vertex]) - set(path):
+            if next == end:
+                yield path + [end]
+            else:
+                stack.append((next, path+[next]))
+
+def get_head(tree):
+
+    """
+    Returns head of the tree.
+
+    Args:
+        tree: dictionary reprenseting binary tree
+        key: is a node, value: (left_child,right_child)
+        for leaves value is ().
+        if right or left child are missing, the value is None
+
+    Returns:
+        head value of the tree
+        None, when tree is empty
+    """
+
+
+    if not tree:
+        return None
+    all_nodes = set()
+    for node in tree.keys():
+        if node:
+            all_nodes.update(list(tree[node]))
+    head = (set(tree.keys())-all_nodes-set([None])).pop()
+    return head
+
+def get_ancestors(tree, node):
+
+    """Returns the ancestors from head to a given node
+    """
+
+    head = get_head(tree)
+    if not head:
+        logging.warning('tree is empty, no ancestors')
+        return None
+    if head == node:
+        logging.warning('Head has no ancestors')
+        return None
+    ancestors = list(dfs_path(tree, head, node, path=None))
```

Comment:
It looks a bit awkward to have to explicitly `list()` the return value from the generator and then also following it up with `ancestors[0][:-1]` below. It might not be appropriate to change the `dfs_path()` return value but I'd suggest at least considering it if a fair amount of postprocessing is required in the caller.